sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,742 @@
+              kafkaSourceDescription.getString(KafkaSourceDescriptionSchemas.TOPIC),
+              kafkaSourceDescription.getInt32(KafkaSourceDescriptionSchemas.PARTITION));
+      Map<String, Object> updatedConsumerConfig =
+          overrideBootstrapServersConfig(consumerConfig, kafkaSourceDescription);
+      try (Consumer<byte[], byte[]> consumer = consumerFactoryFn.apply(updatedConsumerConfig)) {
+        consumerSpEL.evaluateAssign(consumer, ImmutableList.of(topicPartition));
+        long startOffset = tracker.currentRestriction().getFrom();
+        long expectedOffset = startOffset;
+        consumer.seek(topicPartition, startOffset);
+        ConsumerRecords<byte[], byte[]> rawRecords = ConsumerRecords.empty();
+
+        try {
+          while (true) {
+            rawRecords = consumer.poll(kafkaPollTimeout.getMillis());
+            // When there is no records from the current TopicPartition temporarily, self-checkpoint
+            // and move to process the next element.
+            if (rawRecords.isEmpty()) {
+              return ProcessContinuation.resume();
+            }
+            for (ConsumerRecord<byte[], byte[]> rawRecord : rawRecords) {
+              if (!tracker.tryClaim(rawRecord.offset())) {
+                return ProcessContinuation.stop();
+              }
+              KafkaRecord<K, V> kafkaRecord =
+                  new KafkaRecord<>(
+                      rawRecord.topic(),
+                      rawRecord.partition(),
+                      rawRecord.offset(),
+                      consumerSpEL.getRecordTimestamp(rawRecord),
+                      consumerSpEL.getRecordTimestampType(rawRecord),
+                      ConsumerSpEL.hasHeaders() ? rawRecord.headers() : null,
+                      keyDeserializerInstance.deserialize(rawRecord.topic(), rawRecord.key()),
+                      valueDeserializerInstance.deserialize(rawRecord.topic(), rawRecord.value()));
+              Instant outputTimestamp = extractOutputTimestampFn.apply(kafkaRecord);
+              int recordSize =
+                  (rawRecord.key() == null ? 0 : rawRecord.key().length)
+                      + (rawRecord.value() == null ? 0 : rawRecord.value().length);
+              avgRecordSize.update(recordSize);
+              avgOffsetGap.update(expectedOffset - rawRecord.offset());
+              expectedOffset = rawRecord.offset() + 1;
+              receiver.outputWithTimestamp(kafkaRecord, outputTimestamp);
+            }
+          }
+        } catch (Exception anyException) {
+          LOG.error("{}: Exception while reading from Kafka", this, anyException);
+          throw anyException;
+        }
+      }
+    }
+
+    @GetRestrictionCoder
+    public Coder<OffsetRange> restrictionCoder() {
+      return new OffsetRange.Coder();
+    }
+
+    @Setup
+    public void setup() throws Exception {
+      // Start to track record size and offset gap per bundle.
+      avgRecordSize = new KafkaIOUtils.MovingAvg();
```

Comment:
Will `avgRecordSize` / `avgOffsetGap` be consistent across multiple topics / partitions.

If we need to track them at the topic/partition level then they will need to part of the restriction.

Reply:
> Will avgRecordSize / avgOffsetGap be consistent across multiple topics / partitions.

I don't think there is a guarantee . I even don't think there is a guarantee per topic. The `avgRecordSize` and `avgOffsetGap` are used to calculate size in `GetSize`, which should be an estimated one.

Reply:
The UnboundedSource implementation tracked it for a specific TopicPartition and it didn't apply to multiple topics. When we are processing multiple restrictions we'll conflate this information from all of them.

Reply:
Yes, currently `KafkaUnboundedReader` tracks `avgRecordSize` /`avgOffsetGap` per TopicPartition, but I don't think these values are serialized when checkpoint happens. We can bound `avgRecordSize` /`avgOffsetGap` to a TopicPartition by always creating a new `MovingAvg` when `processElement` is started.

Reply:
That makes more sense but relies on knowing the lifecycle of when things get called. What if GetSize is invoked before ProcessElement instead of after. This will lead to implementation assumptions which will prevent SDF evolution.